File: backend/order-service/src/main/java/com/quisin/order/service/model/Order.java
package com.quisin.order.service.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "orders")
@Getter
@Setter
@Builder(toBuilder = true)
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(exclude = "items")
@ToString
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "restaurant_id", nullable = false)
    private Long restaurantId;

    @Column(name = "customer_id", nullable = false)
    private Long customerId;

    @Column(name = "waiter_id")
    private Long waiterId;

    @Column(name = "table_id")
    private Long tableId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderType type;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status;

    @Column(name = "total_amount", precision = 10, scale = 2)
    private BigDecimal totalAmount;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<OrderItem> items = new ArrayList<>();

    @Column(name = "special_instructions")
    private String specialInstructions;

    @Column(name = "group_order_id")
    private Long groupOrderId;

    @Column(name = "is_waiter_assignment_requested")
    private Boolean isWaiterAssignmentRequested;

    @Enumerated(EnumType.STRING)
    @Column(name = "waiter_assignment_status")
    private WaiterAssignmentStatus waiterAssignmentStatus;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    public void addOrderItem(OrderItem item) {
        items.add(item);
        item.setOrder(this);
    }

    public void removeOrderItem(OrderItem item) {
        items.remove(item);
        item.setOrder(null);
    }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (status == null) {
            status = OrderStatus.PENDING;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
} 

File: backend/order-service/src/main/kotlin/com/quisin/order/domain/Order.kt
package com.quisin.order.domain

import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.*

enum class OrderStatus {
    PENDING, 
    CONFIRMED, 
    PREPARING, 
    READY_FOR_PICKUP, 
    COMPLETED, 
    CANCELLED
}

enum class OrderType {
    SINGLE, 
    GROUP
}

enum class WaiterAssignmentStatus {
    PENDING,      // Waiter assignment requested but not yet assigned
    ASSIGNED,     // Waiter has been assigned to the order
    DECLINED,     // Waiter declined the order assignment
    COMPLETED     // Order served by the assigned waiter
}

@Entity
@Table(name = "orders")
@EntityListeners(AuditingEntityListener::class)
data class Order(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID? = null,

    @Column(nullable = false)
    val restaurantId: UUID,

    @Column(nullable = false)
    val customerId: UUID,

    @Column
    val assignedWaiterId: UUID? = null,

    @Column
    val waiterId: UUID? = null,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: OrderStatus = OrderStatus.PENDING,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val type: OrderType,

    @Column(nullable = false, precision = 10, scale = 2)
    val totalAmount: BigDecimal,

    @ElementCollection
    @CollectionTable(name = "order_items", joinColumns = [JoinColumn(name = "order_id")])
    val items: List<OrderItem> = listOf(),

    @Column
    val tableNumber: String? = null,

    @Column
    val specialInstructions: String? = null,

    @Column
    val groupOrderId: UUID? = null,

    @Column(nullable = false)
    val isGroupOrderLeader: Boolean = false,

    @Column(nullable = false)
    val isWaiterAssignmentRequested: Boolean = false,

    @Enumerated(EnumType.STRING)
    @Column
    val waiterAssignmentStatus: WaiterAssignmentStatus? = null,

    @CreatedDate
    @Column(nullable = false, updatable = false)
    val createdAt: LocalDateTime? = null,

    @LastModifiedDate
    @Column(nullable = false)
    val updatedAt: LocalDateTime? = null
)

@Embeddable
data class OrderItem(
    @Column(nullable = false)
    val menuItemId: UUID,

    @Column(nullable = false)
    val quantity: Int,

    @Column(nullable = false, precision = 10, scale = 2)
    val price: BigDecimal,

    @Column
    val specialInstructions: String? = null
)
